Model/Service/Sync.php

```diff
@@ -0,0 +1,261 @@
+     * Index constructor.
+     * @param IndexRepository $indexRepository
+     * @param NostoHelperAccount $nostoHelperAccount
+     * @param NostoHelperUrl $nostoHelperUrl
+     * @param NostoLogger $logger
+     * @param NostoIndexCollectionFactory $nostoIndexCollectionFactory
+     * @param NostoDataHelper $nostoDataHelper
+     */
+    public function __construct(
+        IndexRepository $indexRepository,
+        NostoHelperAccount $nostoHelperAccount,
+        NostoHelperUrl $nostoHelperUrl,
+        NostoLogger $logger,
+        NostoIndexCollectionFactory $nostoIndexCollectionFactory,
+        NostoDataHelper $nostoDataHelper
+    ) {
+        parent::__construct($nostoDataHelper, $logger);
+        $this->indexRepository = $indexRepository;
+        $this->nostoHelperAccount = $nostoHelperAccount;
+        $this->nostoHelperUrl = $nostoHelperUrl;
+        $this->nostoIndexCollectionFactory = $nostoIndexCollectionFactory;
+        $this->nostoDataHelper = $nostoDataHelper;
+    }
+
+    /**
+     * Handles sync of product collection by sending it to Nosto
+     *
+     * @param NostoIndexCollection $collection
+     * @param Store $store
+     * @throws NostoException
+     * @throws MemoryOutOfBoundsException
+     */
+    public function syncIndexedProducts(NostoIndexCollection $collection, Store $store)
+    {
+        if (!$this->nostoDataHelper->isProductUpdatesEnabled($store)) {
+            $this->getLogger()->info(
+                'Nosto product sync is disabled - skipping upserting products to Nosto'
+            );
+            return;
+        }
+        $account = $this->nostoHelperAccount->findAccount($store);
+        if ($account instanceof NostoSignupAccount === false) {
+            throw new NostoException(sprintf('Store view %s does not have Nosto installed', $store->getName()));
+        }
+        $this->startBenchmark(self::BENCHMARK_SYNC_NAME, self::BENCHMARK_SYNC_BREAKPOINT);
+        $collection->setPageSize(self::API_BATCH_SIZE);
+        $iterator = new Iterator($collection);
+        $iterator->eachBatch(function (NostoIndexCollection $collectionBatch) use ($account, $store) {
+            $this->checkMemoryConsumption('product sync');
+            $op = new UpsertProduct($account, $this->nostoHelperUrl->getActiveDomain($store));
+            $op->setResponseTimeout(self::RESPONSE_TIMEOUT);
+            foreach ($collectionBatch as $productIndex) {
+                if (!$productIndex->getInSync()) {
+                    $op->addProduct($productIndex->getNostoProduct());
+                }
+            }
+            try {
+                $op->upsert();
+                $this->tickBenchmark(self::BENCHMARK_SYNC_NAME);
+            } catch (\Exception $upsertException) {
+                $this->getLogger()->exception($upsertException);
+            } finally {
+                // We will set this as in sync even if there was failures
+                $collectionBatch->markAsInSyncCurrentItemsByStore($store);
```

Comment:
Should we simply let the errors bubble to the surface?

Reply:
Let's move this to the try block